packages/core/_src/stm/TQueue/operations/_internal/InternalTQueue.ts

```diff
@@ -0,0 +1,187 @@
+import type { USTM } from "@effect/core/stm/STM/definition"
+import { STMInterruptException, STMRetryException } from "@effect/core/stm/STM/definition"
+import type { Strategy } from "@effect/core/stm/TQueue/definition"
+import { _TIn, _TOut, TQueueSym } from "@effect/core/stm/TQueue/definition"
+
+export class InternalTQueue<A> implements TQueue<A> {
+  readonly [TQueueSym]: TQueueSym = TQueueSym
+  readonly [_TIn]!: (_: A) => void
+  readonly [_TOut]!: () => A
+
+  constructor(
+    readonly ref: TRef<ImmutableQueue<A>>,
+    readonly capacity: number,
+    readonly strategy: Strategy
+  ) {}
+
+  readonly isShutdown: USTM<boolean> = STM.Effect((journal) => {
+    const queue = this.ref.unsafeGet(journal)
+
+    return queue === null
+  })
+
+  offer<A>(a: A): USTM<boolean> {
+    return STM.Effect((journal, fiberId) => {
+      const queue = this.ref.unsafeGet(journal)
+
+      if (queue === null) {
+        throw new STMInterruptException(fiberId)
+      }
+
+      if (queue.size < this.capacity) {
+        this.ref.unsafeSet(queue.append(a), journal)
```

Comment:
@0x706b This is one of the places where it's picking up the `Collection<A>` `append` instead of the `ImmutableQueue<A>` `append` extension.